blog/posts/2024-11-07-how-to-get-a-rust-pkg-to-cran/index.qmd

```diff
@@ -0,0 +1,207 @@
+
+A little bit of foreshadowing, but we will have to edit our `DESCRIPTION` file and add the right level of detail for our package, such as author, description, URLs etc.
+
+
+### 2. Add Rust as a Dependency
+
+Similar to the `usethis` package, there is the `rextendr` package that makes this step pretty straight forward.
+
+```r
+rextendr::use_extendr()
+```
+
+This will create the required files in `src/` and `src/rust`.
+
+As the command tells us, whenever we update our Rust code, we should run the following to document the code and build the Rust-parts.
+
+```r
+rextendr::document()
+# if we have changed our R-code and its documentation
+# we need the following as well
+devtools::document()
+```
+
+And we should be ready to go and call our default Rust function `hello_world()` (defined in `src/rust/src/lib.rs`).
+
+The actual R and Rust functions are typically the easiest parts of developing a package.
+If you need a good starter, have a look here, eg [`R/get_tokens.R`](https://github.com/DavZim/rtiktoken/blob/master/R/get_tokens.R) as well as [`src/rust/src/lib.rs`](https://github.com/DavZim/rtiktoken/blob/master/src/rust/src/lib.rs) (as we can see, I didn't lie when I said it's a *light* wrapper...).
+
+If we need to add a Rust dependency, we can use `rextendr::use_crate()` or use `cargo add xyz` directly from the `src/rust` directory.
+
+Now on to the "hard" parts.
+
+
+### 3. Get the Package to CRAN
+
+First, we need to make sure that the usual hurdles are met, see also the [R Packages (2e) Book](https://r-pkgs.org/).
+
+- document our functions using [`roxygen2`](https://roxygen2.r-lib.org/) and create the documentation using `devtools::document()`
+- fill the details of our `DESCRIPTION` file, write the `README.Rmd` and knit to `README.md`
+- use [`testthat`](https://testthat.r-lib.org/) and write tests (not strictly needed, but will most likely safe us in the future!)
+- ... other steps that are typically done in R package development
+- make sure `devtools::check()` works without a NOTE
+
+There are however a couple of CRAN-specific rules and best practices for packages using Rust (see also [Using Rust in CRAN Packages](https://cran.r-project.org/web/packages/using_rust.html)).
+Most of these requirements are already met, but there are a couple of must-haves and nice-to-haves.
+
+Note that some of the following `rextendr` functions are currently only available in the development version of `rextendr` (>0.3.1).
+
+
+#### CRAN Defaults
+
+First, we should tell `rextendr`, that we want to use the CRAN standards.
+For example, `Makevars` for different platforms, etc.
+We achieve this by calling
+
+```r
+rextendr::use_cran_defaults()
+```
+
+
+#### MSRV
+
+Then, we should find and record our MSRV (Minimal Supported Rust Version).
+Luckily, there is the [`cargo-msrv`](https://github.com/foresterre/cargo-msrv) crate, which tells us what our MSRV is.
+To find our MSRV, we can do the following (from the terminal and not from R this time):
```

Comment:
Perhaps we can use this as a time to educate on the MSRV and how it is found. Something like so: 
```suggestion
Then, we should find and record our MSRV (Minimal Supported Rust Version). The MSRV is the minimum required 
version of rust to be able to build the R package from source. Discovering the MSRV isn't entirely straightforward.
Luckily, there is the [`cargo-msrv`](https://github.com/foresterre/cargo-msrv) crate, which tells us what our MSRV is. 
Finding the MSRV involves compiling the rust source code using different versions of Rust. 
To find our MSRV, we can do the following (from the terminal and not from R this time):
```